### 과제 (Q1 ~ Q10)

Q1. titanic 데이터셋에서 age, fare 2개 열을 선택하여 데이터프레임 df를 만드세요

In [203]:
import pandas as pd
import seaborn as sns

df = sns.load_dataset('titanic')
df.head()

survived  pclass     sex   age  sibsp  parch     fare embarked  class  \
0         0       3    male  22.0      1      0   7.2500        S  Third   
1         1       1  female  38.0      1      0  71.2833        C  First   
2         1       3  female  26.0      0      0   7.9250        S  Third   
3         1       1  female  35.0      1      0  53.1000        S  First   
4         0       3    male  35.0      0      0   8.0500        S  Third   

     who  adult_male deck  embark_town alive  alone  
0    man        True  NaN  Southampton    no  False  
1  woman       False    C    Cherbourg   yes  False  
2  woman       False  NaN  Southampton   yes   True  
3  woman       False    C  Southampton   yes  False  
4    man        True  NaN  Southampton    no   True

In [204]:
df1=df[['age','fare']]
df1.head()

age     fare
0  22.0   7.2500
1  38.0  71.2833
2  26.0   7.9250
3  35.0  53.1000
4  35.0   8.0500

Q2. 10을 더하는 함수 add_10 과 두 객체의 합 add_two_obj 함수를 정의한 후 결과값이 각각 20과 20을 출력하세요.

In [205]:
def add_10(a):
    return a +10

def add_two_obj(a,b):
    return a+b

print(add_10(10))
print(add_two_obj(10,10))

20
20


Q3. 10을 곱하는 사용자 함수와 두 객체의 나눗셈을 수행하는 사용자 함수를 정의한 후 결과값이 각각 200과 10을 출력하세요.

In [206]:
def mul_10(a):
    return a * 10

def divide(a,b):
    return a / b

print(mul_10(20))
print(divide(200,20))

200
10.0


Q4. df.age 에 add_10 함수를 적용해서 출력하세요.

In [207]:
sr1 = df.age.apply(add_10)
sr1

0      32.0
1      48.0
2      36.0
3      45.0
4      45.0
       ... 
886    37.0
887    29.0
888     NaN
889    36.0
890    42.0
Name: age, Length: 891, dtype: float64

Q5. 시리즈 객체 + 숫자를 연산하는 방식으로 df.age에 10을 더한 후 sr2에 저장하고 출력하세요.


In [139]:
sr2 = df.age.apply(add_two_obj, b=10)
sr2

0      32.0
1      48.0
2      36.0
3      45.0
4      45.0
       ... 
886    37.0
887    29.0
888     NaN
889    36.0
890    42.0
Name: age, Length: 891, dtype: float64

Q6. 람다 함수를 활용해서 위와 같은 연산을 수행하고 결과를 sr3로 출력하세요


In [208]:
sr3 = df.age.apply(lambda a:add_10(a))
sr3

0      32.0
1      48.0
2      36.0
3      45.0
4      45.0
       ... 
886    37.0
887    29.0
888     NaN
889    36.0
890    42.0
Name: age, Length: 891, dtype: float64

Q7. df의 'age', 'fare' 칼럼에 10을 덧셈연산하는 사용자 함수를 정의하고 applymap()으로 각 사용자 함수를 매핑한 후 출력하세요.


In [210]:
df_map = df[['age','fare']].applymap(add_10).head()

In [209]:
#df[['age','fare']].applymap(add_10).head()
df_map = df1.applymap(add_10)
df_map.head()

age     fare
0  32.0  17.2500
1  48.0  81.2833
2  36.0  17.9250
3  45.0  63.1000
4  45.0  18.0500

Q8. df_map의 [0,0]을 NaN으로 변경한 후 missing_value 함수를 적용해서 불린 시리즈를 반환하세요.

 - df_map은 7번 문제에서 df_map으로 이름을 사용하신 후 진행하시면 됩니다.


In [211]:
import numpy as np

df_map.iloc[0,0] = np.nan
df_map.head()

age     fare
0   NaN  17.2500
1  48.0  81.2833
2  36.0  17.9250
3  45.0  63.1000
4  45.0  18.0500

In [212]:
def missing_value(series):
    return series.isnull()

df_map= df_map.apply(missing_value, axis=0)
df_map.head()

age   fare
0   True  False
1  False  False
2  False  False
3  False  False
4  False  False

Q9. titanic 데이터셋을 seaborn에서 로딩한 후 블린 인덱싱을 이용하여 아래 사항들을 수행하세요.
- 나이가 10대(10~19세)인 승객만 따로 선택
- 나이가 10세 미만(0~9세)이고 여성인 승객만 따로 선택
- 나이가 10세 미만(0~9세) 또는 60세 이상인 승객의 age, sex, alone 열만 선택


In [213]:
import warnings
warnings.filterwarnings('ignore')

In [214]:
import seaborn as sns
df = sns.load_dataset('titanic')
df.head()

survived  pclass     sex   age  sibsp  parch     fare embarked  class  \
0         0       3    male  22.0      1      0   7.2500        S  Third   
1         1       1  female  38.0      1      0  71.2833        C  First   
2         1       3  female  26.0      0      0   7.9250        S  Third   
3         1       1  female  35.0      1      0  53.1000        S  First   
4         0       3    male  35.0      0      0   8.0500        S  Third   

     who  adult_male deck  embark_town alive  alone  
0    man        True  NaN  Southampton    no  False  
1  woman       False    C    Cherbourg   yes  False  
2  woman       False  NaN  Southampton   yes   True  
3  woman       False    C  Southampton   yes  False  
4    man        True  NaN  Southampton    no   True

In [218]:
# 나이가 10대(10~19세)인 승객만 따로 선택
df.loc[(df['age']>=10) & (df['age']<20),:].head()

survived  pclass     sex   age  sibsp  parch      fare embarked   class  \
9          1       2  female  14.0      1      0   30.0708        C  Second   
14         0       3  female  14.0      0      0    7.8542        S   Third   
22         1       3  female  15.0      0      0    8.0292        Q   Third   
27         0       1    male  19.0      3      2  263.0000        S   First   
38         0       3  female  18.0      2      0   18.0000        S   Third   

      who  adult_male deck  embark_town alive  alone  
9   child       False  NaN    Cherbourg   yes  False  
14  child       False  NaN  Southampton    no   True  
22  child       False  NaN   Queenstown   yes   True  
27    man        True    C  Southampton    no  False  
38  woman       False  NaN  Southampton    no  False

In [219]:
# 나이가 10세 미만(0~9세)이고 여성인 승객만 따로 선택
df.loc[(df['age']<10) & (df['sex']=='female'),:].head()

#mask1 = (df['age']>=10)&(df['age']<20)
#df = df.loc[mask1,:]

survived  pclass     sex  age  sibsp  parch     fare embarked   class  \
10          1       3  female  4.0      1      1  16.7000        S   Third   
24          0       3  female  8.0      3      1  21.0750        S   Third   
43          1       2  female  3.0      1      2  41.5792        C  Second   
58          1       2  female  5.0      1      2  27.7500        S  Second   
119         0       3  female  2.0      4      2  31.2750        S   Third   

       who  adult_male deck  embark_town alive  alone  
10   child       False    G  Southampton   yes  False  
24   child       False  NaN  Southampton    no  False  
43   child       False  NaN    Cherbourg   yes  False  
58   child       False  NaN  Southampton   yes  False  
119  child       False  NaN  Southampton    no  False

In [220]:
# 나이가 10세 미만(0~9세) 또는 60세 이상인 승객의 age, sex, alone 열만 선택
df.loc[(df['age']<10) | (df['age']>=60),['age','sex','alone']].head()

#mask2 = (df['age']<10) | (df['age']>=60)
#df = df.loc[mask2,:]

age     sex  alone
7    2.0    male  False
10   4.0  female  False
16   2.0    male  False
24   8.0  female  False
33  66.0    male   True

Q10. 나이가 10세 미만(0~9세) 또는 60세 이상인 승객의 alone 열만 선택한 후 값을 True로 변경하세요



In [150]:
df.loc[(df['age']<10)|(df['age']>=60),"alone"]=True

In [151]:
df.loc[(df['age']<10)|(df['age']>=60),["alone"]]

alone
7     True
10    True
16    True
24    True
33    True
..     ...
831   True
850   True
851   True
852   True
869   True

[88 rows x 1 columns]

## 그룹화 (groupping)

### 그룹 연산
- 복잡한 데이터를 어떤 기준에 따라 여러 그룹으로 나누어서 관찰할 수 있으며 이런 방식으로 분할 처리하는 것을 그룹 연산이라 함.
- 그룹 연산은 데이터를 집계, 반환, 필터하는데 효율적이며 판다스 groupby() 메소드를 사용
- 그룹 객체 만들기(분할)
- 그룹 연산 메소드(적용-결합)

In [152]:
import pandas as pd
import seaborn as sns #seaborn 라이브러리 불러오기 (dataset)

titanic = sns.load_dataset('titanic')
titanic.columns

Index(['survived', 'pclass', 'sex', 'age', 'sibsp', 'parch', 'fare',
       'embarked', 'class', 'who', 'adult_male', 'deck', 'embark_town',
       'alive', 'alone'],
      dtype='object')

In [153]:
titanic.head()

survived  pclass     sex   age  sibsp  parch     fare embarked  class  \
0         0       3    male  22.0      1      0   7.2500        S  Third   
1         1       1  female  38.0      1      0  71.2833        C  First   
2         1       3  female  26.0      0      0   7.9250        S  Third   
3         1       1  female  35.0      1      0  53.1000        S  First   
4         0       3    male  35.0      0      0   8.0500        S  Third   

     who  adult_male deck  embark_town alive  alone  
0    man        True  NaN  Southampton    no  False  
1  woman       False    C    Cherbourg   yes  False  
2  woman       False  NaN  Southampton   yes   True  
3  woman       False    C  Southampton   yes  False  
4    man        True  NaN  Southampton    no   True

In [231]:
import pandas as pd
import seaborn as sns

titanic = sns.load_dataset('titanic')
df = titanic.loc[:,['age','sex','class','fare','survived']]

print('승객 수:', len(df))
print(df.head())

승객 수: 891
    age     sex  class     fare  survived
0  22.0    male  Third   7.2500         0
1  38.0  female  First  71.2833         1
2  26.0  female  Third   7.9250         1
3  35.0  female  First  53.1000         1
4  35.0    male  Third   8.0500         0


In [232]:
df.columns

Index(['age', 'sex', 'class', 'fare', 'survived'], dtype='object')

#### 그룹 객체 만들기 (분할) group by

In [233]:
# class열을 기준으로 그룹핑(분할)
grouped = df.groupby(['class'])
print(grouped)
list(grouped)

[('First',
        age     sex  class     fare  survived
  1    38.0  female  First  71.2833         1
  3    35.0  female  First  53.1000         1
  6    54.0    male  First  51.8625         0
  11   58.0  female  First  26.5500         1
  23   28.0    male  First  35.5000         1
  ..    ...     ...    ...      ...       ...
  871  47.0  female  First  52.5542         1
  872  33.0    male  First   5.0000         0
  879  56.0  female  First  83.1583         1
  887  19.0  female  First  30.0000         1
  889  26.0    male  First  30.0000         1
  
  [216 rows x 5 columns]),
 ('Second',
        age     sex   class     fare  survived
  9    14.0  female  Second  30.0708         1
  15   55.0  female  Second  16.0000         1
  17    NaN    male  Second  13.0000         1
  20   35.0    male  Second  26.0000         0
  21   34.0    male  Second  13.0000         1
  ..    ...     ...     ...      ...       ...
  866  27.0  female  Second  13.8583         1
  874  28.0  female

In [234]:
# 그룹 객체를 iteration으로 출력
for key, group in grouped:
    print('* key :', key)
    print('* number :',len(group))
    print(group.head())

* key : First
* number : 216
     age     sex  class     fare  survived
1   38.0  female  First  71.2833         1
3   35.0  female  First  53.1000         1
6   54.0    male  First  51.8625         0
11  58.0  female  First  26.5500         1
23  28.0    male  First  35.5000         1
* key : Second
* number : 184
     age     sex   class     fare  survived
9   14.0  female  Second  30.0708         1
15  55.0  female  Second  16.0000         1
17   NaN    male  Second  13.0000         1
20  35.0    male  Second  26.0000         0
21  34.0    male  Second  13.0000         1
* key : Third
* number : 491
    age     sex  class     fare  survived
0  22.0    male  Third   7.2500         0
2  26.0  female  Third   7.9250         1
4  35.0    male  Third   8.0500         0
5   NaN    male  Third   8.4583         0
7   2.0    male  Third  21.0750         0


#### groupping  함수

In [235]:
# 연산 메서드 적용
average = grouped.mean() # mean 평균값 구하기 
average

age       fare  survived
class                                 
First   38.233441  84.154687  0.629630
Second  29.877630  20.662183  0.472826
Third   25.140620  13.675550  0.242363

In [236]:
stat = grouped.max() # max 최댓값 구하기
stat

age   sex      fare  survived
class                                 
First   80.0  male  512.3292         1
Second  70.0  male   73.5000         1
Third   74.0  male   69.5500         1

In [237]:
# Q. 3rd 그룹만을 선택해서 group3 이름으로 저장하고 통계요약표를 출력하세요.
group3 = grouped.get_group('Third') # 3rd 그룹 group3로 저장하기
group3.describe() # describe 각 column별로 다양한 정보 한번에 보기

age        fare    survived
count  355.000000  491.000000  491.000000
mean    25.140620   13.675550    0.242363
std     12.495398   11.778142    0.428949
min      0.420000    0.000000    0.000000
25%     18.000000    7.750000    0.000000
50%     24.000000    8.050000    0.000000
75%     32.000000   15.500000    0.000000
max     74.000000   69.550000    1.000000

In [239]:
# 2개의 column지정해서 groupby하기

grouped_two = df.groupby(['class','sex'])
average_two = grouped_two.mean()
average_two

age        fare  survived
class  sex                                    
First  female  34.611765  106.125798  0.968085
       male    41.281386   67.226127  0.368852
Second female  28.722973   21.970121  0.921053
       male    30.740707   19.741782  0.157407
Third  female  21.750000   16.118810  0.500000
       male    26.507589   12.661633  0.135447

In [240]:
#  get_group 그룹화한 해당 행만 선택하여 보기 / 특정 데이터 추출
# get_group() : 기준으로 한 열의 값을 가진 그룹만을 추출해 낼 수 있다.
group3f = grouped_two.get_group(('Third','female'))
group3f.head()

age     sex  class     fare  survived
2   26.0  female  Third   7.9250         1
8   27.0  female  Third  11.1333         1
10   4.0  female  Third  16.7000         1
14  14.0  female  Third   7.8542         0
18  31.0  female  Third  18.0000         0

#### 그룹 연산 메소드 (적용-결합)

In [241]:
# 모든 열에 여러 함수를 매핑
agg_all = grouped.agg(['min','max'])
agg_all.head()

age           sex       fare           survived    
         min   max     min   max  min       max      min max
class                                                       
First   0.92  80.0  female  male  0.0  512.3292        0   1
Second  0.67  70.0  female  male  0.0   73.5000        0   1
Third   0.42  74.0  female  male  0.0   69.5500        0   1

In [242]:
# 각 열마다 다른 함수를 적용하여 집계
# 그룹 객체에 함수를 매핑할 경우엔 agg()메소드를 사용한다.
agg_sep = grouped.agg({'fare':['min','max'],'age':'mean'})
agg_sep.head()

fare                  age
        min       max       mean
class                           
First   0.0  512.3292  38.233441
Second  0.0   73.5000  29.877630
Third   0.0   69.5500  25.140620

In [243]:
# Q. 데이터 개수가 200개 이상인 그룹만을 필터링하여 데이터프레임으로 반환하여 grouped_filter 이름으로 저장 후
# 처음 5개 행을 출력하세요


grouped_filter = grouped.filter(lambda x:len(x) >=200)
grouped_filter.head()

age     sex  class     fare  survived
0  22.0    male  Third   7.2500         0
1  38.0  female  First  71.2833         1
2  26.0  female  Third   7.9250         1
3  35.0  female  First  53.1000         1
4  35.0    male  Third   8.0500         0

In [244]:
grouped_filter['class'].value_counts()

Third     491
First     216
Second      0
Name: class, dtype: int64

### 과제(7/1)
age열의 평균이 30보다 작은 그룹만을 필터링하여 데이터프레임으로 반환하여 age_filter 이름으로 저장 후 처음 5개 행을 출력하세요.

In [304]:
import pandas as pd
import seaborn as sns

titanic = sns.load_dataset('titanic')
df = titanic.loc[:,['age','sex','class','fare','survived']]

age_filter = grouped.filter(lambda x : x.age.mean() < 30 )
age_filter.head()

age     sex  class     fare  survived
0  22.0    male  Third   7.2500         0
2  26.0  female  Third   7.9250         1
4  35.0    male  Third   8.0500         0
5   NaN    male  Third   8.4583         0
7   2.0    male  Third  21.0750         0

In [301]:
age_filter = grouped.apply(lambda x : x.age.mean() < 30)

for x in age_filter.index:
    if age_filter[x]==True:
        age_filter_df = grouped.get_group(x)
        print(age_filter_df.head())
        print('\n')

     age     sex   class     fare  survived
9   14.0  female  Second  30.0708         1
15  55.0  female  Second  16.0000         1
17   NaN    male  Second  13.0000         1
20  35.0    male  Second  26.0000         0
21  34.0    male  Second  13.0000         1


    age     sex  class     fare  survived
0  22.0    male  Third   7.2500         0
2  26.0  female  Third   7.9250         1
4  35.0    male  Third   8.0500         0
5   NaN    male  Third   8.4583         0
7   2.0    male  Third  21.0750         0




### 피벗 테이블 (pivot)
피벗 테이블을 구성하는 4가지 요소 (행인덱스, 열인덱스, 데이터 값, 데이터 집계함수)에 적용할 데이터프레임의 열을 각각 지정하여 함수의 인자로 전달

In [199]:
import pandas as pd
import seaborn as sns
titanic = sns.load_dataset('titanic')
df = titanic.loc[:,['age','sex','class','fare','survived']]
df.head()

age     sex  class     fare  survived
0  22.0    male  Third   7.2500         0
1  38.0  female  First  71.2833         1
2  26.0  female  Third   7.9250         1
3  35.0  female  First  53.1000         1
4  35.0    male  Third   8.0500         0

In [200]:
pdf1 = pd.pivot_table(df,
                     index = 'class',  # 행인덱스
                     columns = 'sex',   # 열인덱스
                     values = 'fare',   # 데이터 값
                      aggfunc = 'mean')   #데이터 집계함수
pdf1.head()

sex         female       male
class                        
First   106.125798  67.226127
Second   21.970121  19.741782
Third    16.118810  12.661633

In [201]:
# 행, 열, 값에 사용할 열을 2개 이상 지정 - index, values, aggfunc에 각각 2개 이상 지정
pdf2 = pd.pivot_table(df,
                     index=['class','sex'],
                     columns='survived',
                     values=['age','fare'],
                     aggfunc=['mean','max'])
pdf2.head()

mean                                      max        \
                     age                   fare               age         
survived               0          1           0           1     0     1   
class  sex                                                                
First  female  25.666667  34.939024  110.604167  105.978159  50.0  63.0   
       male    44.581967  36.248000   62.894910   74.637320  71.0  80.0   
Second female  36.000000  28.080882   18.250000   22.288989  57.0  55.0   
       male    33.369048  16.022000   19.488965   21.095100  70.0  62.0   
Third  female  23.818182  19.329787   19.773093   12.464526  48.0  63.0   

                                 
                 fare            
survived            0         1  
class  sex                       
First  female  151.55  512.3292  
       male    263.00  512.3292  
Second female   26.00   65.0000  
       male     73.50   39.0000  
Third  female   69.55   31.3875

## 누락 데이터 처리(결측값)

### 누락 데이터 확인

In [305]:
# 누락 데이터의 총 갯수
df_t = sns.load_dataset('titanic')
df_t.isnull().sum()

survived         0
pclass           0
sex              0
age            177
sibsp            0
parch            0
fare             0
embarked         2
class            0
who              0
adult_male       0
deck           688
embark_town      2
alive            0
alone            0
dtype: int64

In [203]:
df_t.embarked.value_counts() #True 값 count

S    644
C    168
Q     77
Name: embarked, dtype: int64

### 누락 데이터 치환

#### fillna() 
 -- 결측값을 원하는 값으로 변경하는 메서드
- df.fillna(값) 
   - 동일한 하나의 값을 넣을 경우
- df.fillna({'컬럼명1':값1,'컬럼명2':값2})
  - 컬럼별로 다른 값을 입력하는 경우

In [204]:
# 누락데이터의 앞의 값으로 대체
df_t.fillna(method='ffill',inplace=True) # ffill 결측값이 바로 위값과 동일하게 설정
df_t.embarked.value_counts()

S    644
C    169
Q     78
Name: embarked, dtype: int64

### 과제(7/1)
- age 컬럼의 결측치를 적절한 값으로 대체하시고 그 근거를 설명하세요.
- deck 컬럼의 삭제 여부를 판단하는 근거를 설명하세요.

In [ ]:
# 참고
df.dropna(subset=['column'],inplace=True)
df.dropna(inplace=True)
df.column.replace(np.nan, df.column.mean(),inplace=True)

#deck : 선실 고유번호의 앞자리 알파벳 (a-z)

In [222]:
import pandas as pd
import seaborn as sns
df = sns.load_dataset('titanic')
df.head()

survived  pclass     sex   age  sibsp  parch     fare embarked  class  \
0         0       3    male  22.0      1      0   7.2500        S  Third   
1         1       1  female  38.0      1      0  71.2833        C  First   
2         1       3  female  26.0      0      0   7.9250        S  Third   
3         1       1  female  35.0      1      0  53.1000        S  First   
4         0       3    male  35.0      0      0   8.0500        S  Third   

     who  adult_male deck  embark_town alive  alone  
0    man        True  NaN  Southampton    no  False  
1  woman       False    C    Cherbourg   yes  False  
2  woman       False  NaN  Southampton   yes   True  
3  woman       False    C  Southampton   yes  False  
4    man        True  NaN  Southampton    no   True

In [223]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 15 columns):
 #   Column       Non-Null Count  Dtype   
---  ------       --------------  -----   
 0   survived     891 non-null    int64   
 1   pclass       891 non-null    int64   
 2   sex          891 non-null    object  
 3   age          714 non-null    float64 
 4   sibsp        891 non-null    int64   
 5   parch        891 non-null    int64   
 6   fare         891 non-null    float64 
 7   embarked     889 non-null    object  
 8   class        891 non-null    category
 9   who          891 non-null    object  
 10  adult_male   891 non-null    bool    
 11  deck         203 non-null    category
 12  embark_town  889 non-null    object  
 13  alive        891 non-null    object  
 14  alone        891 non-null    bool    
dtypes: bool(2), category(2), float64(2), int64(4), object(5)
memory usage: 80.7+ KB


In [225]:
# age 데이터의 평균으로 변경 (NaN 값 제외)
mean_age = df['age'].mean(axis=0) 
df['age'].fillna(mean_age, inplace=True)

In [226]:
df.isnull().sum()

survived         0
pclass           0
sex              0
age              0
sibsp            0
parch            0
fare             0
embarked         2
class            0
who              0
adult_male       0
deck           688
embark_town      2
alive            0
alone            0
dtype: int64

### 누락 데이터 제거

In [227]:
# deck 열의 nan 개수 계산
nan_deck = df['deck'].value_counts(dropna=False)
print(nan_deck)

NaN    688
C       59
B       47
D       33
E       32
A       15
F       13
G        4
Name: deck, dtype: int64


In [228]:
# 891개 중 688개의 NaN값으로 deck 열삭제
df.dropna(axis=1,inplace=True)

In [230]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column      Non-Null Count  Dtype   
---  ------      --------------  -----   
 0   survived    891 non-null    int64   
 1   pclass      891 non-null    int64   
 2   sex         891 non-null    object  
 3   age         891 non-null    float64 
 4   sibsp       891 non-null    int64   
 5   parch       891 non-null    int64   
 6   fare        891 non-null    float64 
 7   class       891 non-null    category
 8   who         891 non-null    object  
 9   adult_male  891 non-null    bool    
 10  alive       891 non-null    object  
 11  alone       891 non-null    bool    
dtypes: bool(2), category(1), float64(2), int64(4), object(3)
memory usage: 65.5+ KB


<img src = 'dataset/apply_applymap_pipe.jpg' STYLE='width:800px;'>

[pipe vs applymap vs apply]  

pipe와 applymap과의 차이는 반환하는 결과가 pipe가 더 다양하다는 것으로 applymap은 데이터프레임 원소에 적용하므로 데이터프레임으로만 리턴되지만 pipe는 사용자 함수의 리턴값에 따라 개별 값, 시리즈, 데이터프레임을 반환할 수 있다.  

현재 가장 많이 쓰이는 메서드는 apply인데 그 이유는 다양한 형태로 값을 입력 받을 수 있고, 반환하는 형태도 폭이 넓기 때문. 결과는 같아보이지만, 시리즈를 이용해 데이터프레임의 행과 열도 이용할 수 있으며 시리즈 그 자체도 함수로 활용할 수 있다. 

만약 데이터프레임에 함수를 적용해야한다면 리턴하는 형태가 많은 pipe 사용도 고려


In [175]:
import pandas as pd
df = pd.DataFrame({
    'col1':[30,40,50,60],
    'col2':[23,35,65,45],
    'col3':[85,87,90,89],   
}, index = list('ABCD'))
df

col1  col2  col3
A    30    23    85
B    40    35    87
C    50    65    90
D    60    45    89

In [176]:
df.apply(lambda x:x.sum(),axis=0)

col1    180
col2    168
col3    351
dtype: int64

In [177]:
df.col1.apply(lambda x:x+1)

A    31
B    41
C    51
D    61
Name: col1, dtype: int64

In [178]:
df.col1.map(lambda x : x+1)

A    31
B    41
C    51
D    61
Name: col1, dtype: int64

In [179]:
df.applymap(lambda x:'%.2f'% x)

col1   col2   col3
A  30.00  23.00  85.00
B  40.00  35.00  87.00
C  50.00  65.00  90.00
D  60.00  45.00  89.00

In [180]:
# The pipe() method allows you to apply one or more functions to the DataFrame object.

import pandas as pd

def change_age(x):
  x["age"]=[10, 20, 30]
  return x

data = {
  "name": ["Sally", "Mary", "John"],
  "age": [50, 40, 30]
}

df = pd.DataFrame(data)
df


name  age
0  Sally   50
1   Mary   40
2   John   30

In [181]:
df.pipe(change_age)

name  age
0  Sally   10
1   Mary   20
2   John   30

### 과제(Q11 ~ Q15)
Q11 ~ Q15 문제를 수행하세요

Q11. 최대값 - 최소값을 반환하는 사용자 함수를 min_max로 정의하고 age, fare 열로 구성되는 타이타닉 데이터프레임df 에 적용하여 결과와 자료형을 출력하세요. source : titanic = sns.load_dataset('titanic')

In [286]:
import pandas as pd
import seaborn as sns

titanic = sns.load_dataset('titanic')
df = titanic[['age','fare']]
df.head()

age     fare
0  22.0   7.2500
1  38.0  71.2833
2  26.0   7.9250
3  35.0  53.1000
4  35.0   8.0500

In [287]:
def min_max(x):
    result = max(x)-min(x)
    return result

df = df.apply(min_max, axis=0)
df.head()

age      79.5800
fare    512.3292
dtype: float64

Q12. df 에서 'age' 칼럼과 'fare'칼럼을 더한 가격을 리턴하는 사용자 함수를 정의한 후 df['add'] 변수를 생성하여 출력하세요.(lambda 함수 사용)


In [222]:
import pandas as pd
import seaborn as sns

titanic = sns.load_dataset('titanic')
df = titanic[['age','fare']]

In [223]:
def adds(a,b):
    return a+b
df["add"] = df.apply(lambda x:adds(x["age"], x["fare"]), axis = 1)
df.head()

age     fare       add
0  22.0   7.2500   29.2500
1  38.0  71.2833  109.2833
2  26.0   7.9250   33.9250
3  35.0  53.1000   88.1000
4  35.0   8.0500   43.0500

In [285]:
df['add'] = df[['age','fare']].apply(lambda series:series.sum(),axis=1)
df.head()

age     fare       add
0  22.0   7.2500   29.2500
1  38.0  71.2833  109.2833
2  26.0   7.9250   33.9250
3  35.0  53.1000   88.1000
4  35.0   8.0500   43.0500

Q13. 'age', 'fare' 두개 칼럼으로 구성되는 titanic 데이터프레임에서 [0,1]의 값을 NaN을 변경한 후 df 이름으로 출력하세요.


In [290]:
import pandas as pd
import seaborn as sns

titanic = sns.load_dataset('titanic')

titanic = titanic[['age','fare']]
titanic.head()

age     fare
0  22.0   7.2500
1  38.0  71.2833
2  26.0   7.9250
3  35.0  53.1000
4  35.0   8.0500

In [292]:
import numpy as np

df = titanic[['age','fare']]
df.iloc[0,1]=np.nan
df.head()

age     fare
0  22.0      NaN
1  38.0  71.2833
2  26.0   7.9250
3  35.0  53.1000
4  35.0   8.0500

Q14. 각 열의 NaN 찾기, 각 열의 NaN 개수 반환, 데이터프레임의 총 NaN 개수를 반환하는 사용자 함수를 정의한 후 df.pipe 메소드로 함수 매핑한 결과를 모두 출력하세요.



In [271]:
import pandas as pd
import seaborn as sns

titanic = sns.load_dataset('titanic')

# 각 열의 NaN 찾기
def missing_value(x):
    return x.isnull()

# 각 열의 NaN 개수 반환
def missing_count(x):
    return missing_value(x).sum()

# 데이터프레임의 총 NaN 개수를 반환
def total_number_missing(x):
    return missing_count(x).sum()


In [272]:
result = titanic.pipe(missing_value)
result

survived  pclass    sex    age  sibsp  parch   fare  embarked  class  \
0       False   False  False  False  False  False  False     False  False   
1       False   False  False  False  False  False  False     False  False   
2       False   False  False  False  False  False  False     False  False   
3       False   False  False  False  False  False  False     False  False   
4       False   False  False  False  False  False  False     False  False   
..        ...     ...    ...    ...    ...    ...    ...       ...    ...   
886     False   False  False  False  False  False  False     False  False   
887     False   False  False  False  False  False  False     False  False   
888     False   False  False   True  False  False  False     False  False   
889     False   False  False  False  False  False  False     False  False   
890     False   False  False  False  False  False  False     False  False   

       who  adult_male   deck  embark_town  alive  alone  
0    False       False   True        False  False  False  
1    False       False  False        False  False  False  
2    False       False   True        False  False  False  
3    False       False  False        False  False  False  
4    False       False   True        False  False  False  
..     ...         ...    ...          ...    ...    ...  
886  False       False   True        False  False  False  
887  False       False  False        False  False  False  
888  False       False   True        False  False  False  
889  False       False  False        False  False  False  
890  False       False   True        False  False  False  

[891 rows x 15 columns]

In [273]:
result2 = titanic.pipe(missing_count)
result2

survived         0
pclass           0
sex              0
age            177
sibsp            0
parch            0
fare             0
embarked         2
class            0
who              0
adult_male       0
deck           688
embark_town      2
alive            0
alone            0
dtype: int64

In [274]:
result3 = titanic.pipe(total_number_missing)
result3

869

Q15. 타이타닉 df의 열 순서를 아래와 같이 재구성하고 출력하세요
* 열 선택 : ['pclass', 'sex', 'age', 'survived'].
* 열 이름의 리스트 만들기
* 열 이름을 알파벳 순으로 정렬하기
* 열 이름을 기존 순서의 정반대 역순으로 정렬하기
* 열 이름을 사용자가 정의한 임의의 순서로 재배치하기


In [83]:
#열 선택 : ['pclass', 'sex', 'age', 'survived'].
import pandas as pd
import seaborn as sns

titanic = sns.load_dataset('titanic')

df = titanic[['pclass', 'sex', 'age', 'survived']]
df.head()

pclass     sex   age  survived
0       3    male  22.0         0
1       1  female  38.0         1
2       3  female  26.0         1
3       1  female  35.0         1
4       3    male  35.0         0

In [96]:
# 열 이름의 리스트 만들기
list(df.columns)

['pclass', 'sex', 'age', 'survived']

In [99]:
# 열 이름을 알파벳 순으로 정렬하기
df.sort_index(axis=1).head()

age  pclass     sex  survived
0  22.0       3    male         0
1  38.0       1  female         1
2  26.0       3  female         1
3  35.0       1  female         1
4  35.0       3    male         0

In [ ]:
# 열 이름을 기존 순서의 정반대 역순으로 정렬하기

In [100]:
df.sort_index(axis=1,ascending=False).head()

survived     sex  pclass   age
0         0    male       3  22.0
1         1  female       1  38.0
2         1  female       3  26.0
3         1  female       1  35.0
4         0    male       3  35.0

In [ ]:
# 열 이름을 사용자가 정의한 임의의 순서로 재배치하기

In [106]:
df = titanic[['sex', 'pclass', 'survived', 'age']]
df.head()

sex  pclass  survived   age
0    male       3         0  22.0
1  female       1         1  38.0
2  female       3         1  26.0
3  female       1         1  35.0
4    male       3         0  35.0

## 중복 데이터 처리

In [12]:
# 중복
df = pd.DataFrame({'cl':['a','a','b','a','b'],
                  'c2':[1,1,1,2,2],
                  'c3':[1,1,2,2,2]})
df

cl  c2  c3
0  a   1   1
1  a   1   1
2  b   1   2
3  a   2   2
4  b   2   2

### 중복 데이터 확인

In [13]:
# 전체 행 데이터중에서 중복
df_dup = df.duplicated()
df_dup

0    False
1     True
2    False
3    False
4    False
dtype: bool

In [14]:
# 특정 열 데이터중에서 중복
col_dup = df.c2.duplicated()
col_dup

0    False
1     True
2     True
3    False
4     True
Name: c2, dtype: bool

### 중복 데이터 제거

In [17]:
# drop_duplicates() : 중복행 제거
df2 = df.drop_duplicates()
df2

cl  c2  c3
0  a   1   1
2  b   1   2
3  a   2   2
4  b   2   2

In [16]:
# 특정 열 기준으로 중복행 제거
df3 = df.drop_duplicates(subset=['c2','c3'])
df3

cl  c2  c3
0  a   1   1
2  b   1   2
3  a   2   2

## 데이터 표준화

### 자료형 변환

In [5]:
# 데이터 단위 변경
auto_df = pd.read_csv('dataset/auto-mpg.csv')
auto_df.head()

mpg  cylinders  displacement horsepower  weight  acceleration  model year  \
0  18.0          8         307.0        130    3504          12.0          70   
1  15.0          8         350.0        165    3693          11.5          70   
2  18.0          8         318.0        150    3436          11.0          70   
3  16.0          8         304.0        150    3433          12.0          70   
4  17.0          8         302.0        140    3449          10.5          70   

   origin                   car name  
0       1  chevrolet chevelle malibu  
1       1          buick skylark 320  
2       1         plymouth satellite  
3       1              amc rebel sst  
4       1                ford torino

In [91]:
# mpg를 kpl로 환산
mpg_to_kpl = 1.60934 / 3.78541
mpg_to_kpl

0.42514285110463595

In [92]:
auto_df['kpl']= round(auto_df['mpg']*mpg_to_kpl,2)
auto_df[['mpg','kpl']].head()

mpg   kpl
0  18.0  7.65
1  15.0  6.38
2  18.0  7.65
3  16.0  6.80
4  17.0  7.23

In [93]:
auto_df.horsepower.unique()

array(['130', '165', '150', '140', '198', '220', '215', '225', '190',
       '170', '160', '95', '97', '85', '88', '46', '87', '90', '113',
       '200', '210', '193', '?', '100', '105', '175', '153', '180', '110',
       '72', '86', '70', '76', '65', '69', '60', '80', '54', '208', '155',
       '112', '92', '145', '137', '158', '167', '94', '107', '230', '49',
       '75', '91', '122', '67', '83', '78', '52', '61', '93', '148',
       '129', '96', '71', '98', '115', '53', '81', '79', '120', '152',
       '102', '108', '68', '58', '149', '89', '63', '48', '66', '139',
       '103', '125', '133', '138', '135', '142', '77', '62', '132', '84',
       '64', '74', '116', '82'], dtype=object)

In [94]:
import numpy as np
auto_df.horsepower.replace('?',np.nan,inplace=True)
auto_df.dropna(subset=['horsepower'],axis=0,inplace=True)
auto_df.horsepower.isnull().sum()

0

In [95]:
auto_df.horsepower = auto_df.horsepower.astype('float')
auto_df.horsepower.dtypes

dtype('float64')

### 과제(7/1)
'?'가 있는 인덱스를 찾아서 삭제하세요.

In [193]:
index_name=auto_df[auto_df['horsepower']=='?'].index

In [194]:
auto_df.drop(index_name)

mpg  cylinders  displacement  horsepower  weight  acceleration  \
0    18.0          8         307.0       130.0    3504          12.0   
1    15.0          8         350.0       165.0    3693          11.5   
2    18.0          8         318.0       150.0    3436          11.0   
3    16.0          8         304.0       150.0    3433          12.0   
4    17.0          8         302.0       140.0    3449          10.5   
..    ...        ...           ...         ...     ...           ...   
393  27.0          4         140.0        86.0    2790          15.6   
394  44.0          4          97.0        52.0    2130          24.6   
395  32.0          4         135.0        84.0    2295          11.6   
396  28.0          4         120.0        79.0    2625          18.6   
397  31.0          4         119.0        82.0    2720          19.4   

     model year  origin                   car name    kpl  
0            70       1  chevrolet chevelle malibu   7.65  
1            70       1          buick skylark 320   6.38  
2            70       1         plymouth satellite   7.65  
3            70       1              amc rebel sst   6.80  
4            70       1                ford torino   7.23  
..          ...     ...                        ...    ...  
393          82       1            ford mustang gl  11.48  
394          82       2                  vw pickup  18.71  
395          82       1              dodge rampage  13.60  
396          82       1                ford ranger  11.90  
397          82       1                 chevy s-10  13.18  

[392 rows x 10 columns]

In [195]:
auto_df.horsepower.unique()

array([130., 165., 150., 140., 198., 220., 215., 225., 190., 170., 160.,
        95.,  97.,  85.,  88.,  46.,  87.,  90., 113., 200., 210., 193.,
       100., 105., 175., 153., 180., 110.,  72.,  86.,  70.,  76.,  65.,
        69.,  60.,  80.,  54., 208., 155., 112.,  92., 145., 137., 158.,
       167.,  94., 107., 230.,  49.,  75.,  91., 122.,  67.,  83.,  78.,
        52.,  61.,  93., 148., 129.,  96.,  71.,  98., 115.,  53.,  81.,
        79., 120., 152., 102., 108.,  68.,  58., 149.,  89.,  63.,  48.,
        66., 139., 103., 125., 133., 138., 135., 142.,  77.,  62., 132.,
        84.,  64.,  74., 116.,  82.])

Q. 아래 사항을 처리 하세요.
* origin 열의 고유값을 출력하세요.
* 정수형 데이터를 문자형 데이터로 변환한 후 고유값을 출력하세요.(1:'USA', 2:'EU', 3:'JAPAN')


In [17]:
auto_df = pd.read_csv('dataset/auto-mpg.csv')
auto_df.head()

mpg  cylinders  displacement horsepower  weight  acceleration  model year  \
0  18.0          8         307.0        130    3504          12.0          70   
1  15.0          8         350.0        165    3693          11.5          70   
2  18.0          8         318.0        150    3436          11.0          70   
3  16.0          8         304.0        150    3433          12.0          70   
4  17.0          8         302.0        140    3449          10.5          70   

   origin                   car name  
0       1  chevrolet chevelle malibu  
1       1          buick skylark 320  
2       1         plymouth satellite  
3       1              amc rebel sst  
4       1                ford torino

In [18]:
auto_df.origin.unique() # 고유값 출력

array([1, 3, 2], dtype=int64)

In [19]:
auto_df.origin.replace([1, 2, 3],['USA', 'EU', 'JAPAN'], inplace=True)

In [20]:
auto_df.origin.replace({1:'USA', 2:'EU', 3:'JAPAN'},inplace=True)

In [21]:
auto_df.origin.unique()

array(['USA', 'JAPAN', 'EU'], dtype=object)

In [22]:
auto_df.origin.dtypes

dtype('O')

In [23]:
auto_df.origin = auto_df.origin.astype('category') #카테고리 유형으로 type 변경
auto_df.origin.dtypes

CategoricalDtype(categories=['EU', 'JAPAN', 'USA'], ordered=False)

In [24]:
auto_df.origin = auto_df.origin.astype('str') # 문자형으로 변경 'o' => object
auto_df.origin.dtypes

dtype('O')

In [25]:
auto_df = pd.read_pickle('dataset/auto-mpg.pkl')
auto_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 392 entries, 0 to 397
Data columns (total 9 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   mpg           392 non-null    float64
 1   cylinders     392 non-null    int64  
 2   displacement  392 non-null    float64
 3   horsepower    392 non-null    float64
 4   weight        392 non-null    int64  
 5   acceleration  392 non-null    float64
 6   model year    392 non-null    int64  
 7   origin        392 non-null    int64  
 8   car name      392 non-null    object 
dtypes: float64(4), int64(4), object(1)
memory usage: 30.6+ KB


## 범주형 데이터 처리

### 구간 분할

In [ ]:
cut(df,bins,include_lowest,labe)
# df 데이터프레임
# bins 자를 구간
# include_lowest 최소값을 포함할지를 T/F로 돌려받는다
# label 구간의 이름을 지정해준다

In [ ]:
pandas.cut(
x,                    # 구간을 나누고자 하는 데이터(list)
bins,                 # 나누고자 하는 구간(int, sequenc of scalars, or intervallndex)
right = True,        # # 부등호 포함 위치(bool)
labels = None,       # 구간의 라벨(array or False)
retbins = False,     # bins 반환 여부(bool)
duplicates = 'raise', # bins의 중복값 에러 처리 여부(raise or drop)
)

In [ ]:
np.histogram (구간 별 데이터 개수, 구간이 나눠지는 경계값)을 튜플의 형태로 반환한다.
count,bin_divers = 코드~ 값을 할당했으면 count에는 구간 별 데이터 개수 bin_diviers에는 경계값이 저장될 것이다.

In [27]:
# Q. horsepower를 3개의 bin으로 나누고 bin_names = ['저출력','보통출력','고출력']으로 출력하세요.
# cut 함수 : 실수 값의 경계를 지정하는 함수. 각 범주가 모두 동일한 길이를 갖도록 데이터를 범주화


import numpy as np
count, bin_dividers = np.histogram(auto_df.horsepower,bins=3) 
print(bin_dividers)
bin_names = ['저출력','보통출력','고출력']
auto_df['hp_bin'] = pd.cut(x=auto_df.horsepower, # 데이터 배열 
                          bins=bin_dividers,     # 경계값 리스트
                          labels = bin_names,    # bin 이름
                          include_lowest=True) # 첫 경계깞 포함

auto_df[['horsepower','hp_bin']].head()

[ 46.         107.33333333 168.66666667 230.        ]


horsepower hp_bin
0       130.0   보통출력
1       165.0   보통출력
2       150.0   보통출력
3       150.0   보통출력
4       140.0   보통출력

## 정규화
    데이터 값을 동일한 크기 기준으로 나누는 이 과정을 바로 정규화


### 최대값으로 데이터 값을 나누는 정규화

In [30]:
# Q. horsepower 열의 최대값의 절대값으로 모든 데이터를 나누는 정규화를 수행한 후 통계요약표를 출력하세요.
auto_df.horsepower = auto_df.horsepower / abs(auto_df.horsepower.max())
auto_df[['horsepower']].describe()

horsepower
count  392.000000
mean     0.454215
std      0.167353
min      0.200000
25%      0.326087
50%      0.406522
75%      0.547826
max      1.000000

### 과제(7/4)
Q. height 의 값을 cm로 변환하여 새로운 컬럼'hight_cm'을 생성하세요.

In [31]:
fifa = pd.read_csv('dataset/FiFA.csv',encoding='cp949')
fifa.head()

ID               Name  Age Nationality  Overall                 Club  \
0  158023           L. Messi   31   Argentina       94         FC Barcelona   
1   20801  Cristiano Ronaldo   33    Portugal       94             Juventus   
2  190871          Neymar Jr   26      Brazil       92  Paris Saint-Germain   
3  193080             De Gea   27       Spain       91    Manchester United   
4  192985       K. De Bruyne   27     Belgium       91      Manchester City   

  Preferred_Foot       Work_Rate Position  Jersey_Number  \
0           Left  Medium/ Medium       RF             10   
1          Right       High/ Low       ST              7   
2          Right    High/ Medium       LW             10   
3          Right  Medium/ Medium       GK              1   
4          Right      High/ High      RCM              7   

   Contract_Valid_Until Height  Weight_lb  Release_Clause   Value  Wage  
0                  2021    5'7        159          226500  110500   565  
1                  2022    6'2        183          127100   77000   405  
2                  2022    5'9        150          228100  118500   290  
3                  2020    6'4        168          138600   72000   260  
4                  2023   5'11        154          196400  102000   355

In [ ]:
# 1inch = 2.54cm

In [ ]:
#1feet=30.48cm
#1inch=2.54cm
#함수생성(cm 변경)
def inch_cm(x):
    return int(x[0])*30.48+int(x[1])*2.54

# feet와 inch 문자 나누기 split사용
fifa_height=fifa.Height.apply(lambda x: x.split("'"))
type(fifa_height) #문자타입 확인

#inch_cm함수 대입
fifa['hight_cm']=fifa_height.apply(lambda x: inch_cm(x))
fifa.head()

## 데이터프레임 결합 (join)
* 판다스 join() 메소드는 merge() 함수를 기반으로 만들어졌기 때문에 기본 작동 방식이 서로 비슷
* join() 메소드는 두 데이터프레임의 행 인덱스를 기준으로 결합하는 점에서 merge와 차이가 있으나 on=keys 옵션 설정하면 열을 기준으로 결합


In [33]:
data1 = np.arange(1,21).reshape(4,5)
data2 = np.arange(11,31).reshape(5,4)
df1 = pd.DataFrame(data1, columns=list('abcde'))
df2 = pd.DataFrame(data2, columns=list('dehi'))
print(df1,'\n')
print(df2)

    a   b   c   d   e
0   1   2   3   4   5
1   6   7   8   9  10
2  11  12  13  14  15
3  16  17  18  19  20 

    d   e   h   i
0  11  12  13  14
1  15  16  17  18
2  19  20  21  22
3  23  24  25  26
4  27  28  29  30


In [34]:
df1.join(df2,lsuffix='_a',rsuffix='_b')

a   b   c  d_a  e_a  d_b  e_b   h   i
0   1   2   3    4    5   11   12  13  14
1   6   7   8    9   10   15   16  17  18
2  11  12  13   14   15   19   20  21  22
3  16  17  18   19   20   23   24  25  26

In [35]:
# inner default값 inner 옵션을 적용하여 두 데이터프레임에 공통된 집합만을 사용할 수도 있다
df1.join(df2,lsuffix='_a',rsuffix='_b',how='inner')

a   b   c  d_a  e_a  d_b  e_b   h   i
0   1   2   3    4    5   11   12  13  14
1   6   7   8    9   10   15   16  17  18
2  11  12  13   14   15   19   20  21  22
3  16  17  18   19   20   23   24  25  26

In [36]:
df1.join(df2,lsuffix='_a',rsuffix='_b',how='outer')

a     b     c   d_a   e_a  d_b  e_b   h   i
0   1.0   2.0   3.0   4.0   5.0   11   12  13  14
1   6.0   7.0   8.0   9.0  10.0   15   16  17  18
2  11.0  12.0  13.0  14.0  15.0   19   20  21  22
3  16.0  17.0  18.0  19.0  20.0   23   24  25  26
4   NaN   NaN   NaN   NaN   NaN   27   28  29  30

In [37]:
df1['ind']=[1,2,3,4]
df2['ind']=[1,2,3,4,5]
df1

a   b   c   d   e  ind
0   1   2   3   4   5    1
1   6   7   8   9  10    2
2  11  12  13  14  15    3
3  16  17  18  19  20    4

In [38]:
# 특정 열 index 설정
df1.set_index('ind').join(df2.set_index('ind'),lsuffix='_a',rsuffix='_b',on='ind')

a   b   c  d_a  e_a  d_b  e_b   h   i
ind                                        
1     1   2   3    4    5   11   12  13  14
2     6   7   8    9   10   15   16  17  18
3    11  12  13   14   15   19   20  21  22
4    16  17  18   19   20   23   24  25  26

### 합집합

In [40]:
df_s1 = pd.read_excel('dataset/stock price.xlsx',engine='openpyxl',index_col='id')
df_s2 = pd.read_excel('dataset/stock valuation.xlsx',engine='openpyxl',index_col='id')
print(df_s1,'\n')
print(df_s2)

       stock_name          value   price
id                                      
128940       한미약품   59385.666667  421000
130960     CJ E&M   58540.666667   98900
138250      엔에스쇼핑   14558.666667   13200
139480        이마트  239230.833333  254500
142280     녹십자엠에스     468.833333   10200
145990        삼양사   82750.000000   82000
185750        종근당   40293.666667  100500
192400      쿠쿠홀딩스  179204.666667  177500
199800         툴젠   -2514.333333  115400
204210     모두투어리츠    3093.333333    3475 

             name           eps     bps        per       pbr
id                                                          
130960     CJ E&M   6301.333333   54068  15.695091  1.829178
136480         하림    274.166667    3551  11.489362  0.887074
138040    메리츠금융지주   2122.333333   14894   6.313806  0.899691
139480        이마트  18268.166667  295780  13.931338  0.860437
145990        삼양사   5741.000000  108090  14.283226  0.758627
161390      한국타이어   5648.500000   51341   7.453306  0.820007
181710  NHN엔터테인먼트 

In [41]:
# 오른쪽 df 기준으로 join / 매칭이 되지 않은 컬럼의 값들은 'NaN'으로 출력
df_j1 = df_s1.join(df_s2, how='right')
df_j1

stock_name          value     price       name           eps     bps  \
id                                                                            
130960     CJ E&M   58540.666667   98900.0     CJ E&M   6301.333333   54068   
136480        NaN            NaN       NaN         하림    274.166667    3551   
138040        NaN            NaN       NaN    메리츠금융지주   2122.333333   14894   
139480        이마트  239230.833333  254500.0        이마트  18268.166667  295780   
145990        삼양사   82750.000000   82000.0        삼양사   5741.000000  108090   
161390        NaN            NaN       NaN      한국타이어   5648.500000   51341   
181710        NaN            NaN       NaN  NHN엔터테인먼트   2110.166667   78434   
185750        종근당   40293.666667  100500.0        종근당   3990.333333   40684   
204210     모두투어리츠    3093.333333    3475.0     모두투어리츠     85.166667    5335   
207940        NaN            NaN       NaN   삼성바이오로직스   4644.166667   60099   

              per       pbr  
id                           
130960  15.695091  1.829178  
136480  11.489362  0.887074  
138040   6.313806  0.899691  
139480  13.931338  0.860437  
145990  14.283226  0.758627  
161390   7.453306  0.820007  
181710  30.755864  0.827447  
185750  25.185866  2.470259  
204210  40.802348  0.651359  
207940  89.790059  6.938551

In [42]:
# 왼쪽의 df 기준으로 join / 매칭이 되지 않은 컬럼의 값들은 'NaN'으로 출력
df_j1 = df_s1.join(df_s2, how='left')
df_j1

stock_name          value   price    name           eps       bps  \
id                                                                         
128940       한미약품   59385.666667  421000     NaN           NaN       NaN   
130960     CJ E&M   58540.666667   98900  CJ E&M   6301.333333   54068.0   
138250      엔에스쇼핑   14558.666667   13200     NaN           NaN       NaN   
139480        이마트  239230.833333  254500     이마트  18268.166667  295780.0   
142280     녹십자엠에스     468.833333   10200     NaN           NaN       NaN   
145990        삼양사   82750.000000   82000     삼양사   5741.000000  108090.0   
185750        종근당   40293.666667  100500     종근당   3990.333333   40684.0   
192400      쿠쿠홀딩스  179204.666667  177500     NaN           NaN       NaN   
199800         툴젠   -2514.333333  115400     NaN           NaN       NaN   
204210     모두투어리츠    3093.333333    3475  모두투어리츠     85.166667    5335.0   

              per       pbr  
id                           
128940        NaN       NaN  
130960  15.695091  1.829178  
138250        NaN       NaN  
139480  13.931338  0.860437  
142280        NaN       NaN  
145990  14.283226  0.758627  
185750  25.185866  2.470259  
192400        NaN       NaN  
199800        NaN       NaN  
204210  40.802348  0.651359

### 교집합

In [43]:
# id 열을 기준으로 df1와 df2에 공통으로 존재하는 행들을 기준으로 데이터프레임이 결합됐음을 확인할 수 있다
df_j2 = df_s1.join(df_s2, how='inner')
df_j2

stock_name          value   price    name           eps     bps  \
id                                                                       
130960     CJ E&M   58540.666667   98900  CJ E&M   6301.333333   54068   
139480        이마트  239230.833333  254500     이마트  18268.166667  295780   
145990        삼양사   82750.000000   82000     삼양사   5741.000000  108090   
185750        종근당   40293.666667  100500     종근당   3990.333333   40684   
204210     모두투어리츠    3093.333333    3475  모두투어리츠     85.166667    5335   

              per       pbr  
id                           
130960  15.695091  1.829178  
139480  13.931338  0.860437  
145990  14.283226  0.758627  
185750  25.185866  2.470259  
204210  40.802348  0.651359

### 과제(7/4)
Q df_j2에서 중복된 칼럼을 삭제한 후 블린 인덱싱을 이용하여 eps가 3000보다 적거나 stock_name이 이마트인 데이터를 선택하여
데이터프레임을 생성하고 df_j3이름으로 저장 및 출력하세요.
(단, '<'와'=='를 반드시 사용해야 함)

In [74]:
df_j2.T.duplicated()

stock_name    False
value         False
price         False
name           True
eps           False
bps           False
per           False
pbr           False
dtype: bool

In [75]:
~df_j2.T.duplicated()

stock_name     True
value          True
price          True
name          False
eps            True
bps            True
per            True
pbr            True
dtype: bool

In [76]:
df_j2.loc[:,~df_j2.T.duplicated()]

stock_name          value   price           eps     bps        per  \
id                                                                          
130960     CJ E&M   58540.666667   98900   6301.333333   54068  15.695091   
139480        이마트  239230.833333  254500  18268.166667  295780  13.931338   
145990        삼양사   82750.000000   82000   5741.000000  108090  14.283226   
185750        종근당   40293.666667  100500   3990.333333   40684  25.185866   
204210     모두투어리츠    3093.333333    3475     85.166667    5335  40.802348   

             pbr  
id                
130960  1.829178  
139480  0.860437  
145990  0.758627  
185750  2.470259  
204210  0.651359

In [79]:
df_j3= df_j2[(df_j2['eps']<3000)|(df_j2['stock_name']=='이마트')]
df_j3

stock_name          value   price    name           eps     bps  \
id                                                                       
139480        이마트  239230.833333  254500     이마트  18268.166667  295780   
204210     모두투어리츠    3093.333333    3475  모두투어리츠     85.166667    5335   

              per       pbr  
id                           
139480  13.931338  0.860437  
204210  40.802348  0.651359

## 데이터프레임 병합 (merge)
* merge() 함수는 sql의 join 명령과 비슷한 방식으로 어떤 기준에 의해 두 데이터프레임을 병합하는 개념
* 기준이 되는 열이나 인덱스를 키라고 부름
* 키가 되는 열이나 인덱스는 반드시 양쪽 데이터프레임에 모두 존재함


In [ ]:
# pd.merge(df1,df2,how='왼쪽 또는 오른쪽',on='결합기준')

In [45]:
data1 = np.arange(1,21).reshape(4,5)
data2 = np.arange(11,31).reshape(5,4)
df1 = pd.DataFrame(data1, columns=list('abcde'))
df2 = pd.DataFrame(data2, columns=list('dehi'))
print(df1,'\n')
print(df2)

    a   b   c   d   e
0   1   2   3   4   5
1   6   7   8   9  10
2  11  12  13  14  15
3  16  17  18  19  20 

    d   e   h   i
0  11  12  13  14
1  15  16  17  18
2  19  20  21  22
3  23  24  25  26
4  27  28  29  30


In [46]:
df1['ind']=[1,2,3,4]
df2['ind']=[1,2,3,4,5]
df1

a   b   c   d   e  ind
0   1   2   3   4   5    1
1   6   7   8   9  10    2
2  11  12  13  14  15    3
3  16  17  18  19  20    4

In [47]:
# on 절을 지정해주지 않으면 df1,df2의 컬럼명이 동일값으로 join되며,
# 해당 커럼의 값이 일치하지 않으면 NaN 값으로 출력이 된다.
pd.merge(df1,df2,on='ind',how='outer') # how = 'outer' 합집합

a     b     c   d_x   e_x  ind  d_y  e_y   h   i
0   1.0   2.0   3.0   4.0   5.0    1   11   12  13  14
1   6.0   7.0   8.0   9.0  10.0    2   15   16  17  18
2  11.0  12.0  13.0  14.0  15.0    3   19   20  21  22
3  16.0  17.0  18.0  19.0  20.0    4   23   24  25  26
4   NaN   NaN   NaN   NaN   NaN    5   27   28  29  30

In [48]:
df1.count()

a      4
b      4
c      4
d      4
e      4
ind    4
dtype: int64

### 과제(7/4)

df1, df2에 대하여 다음을 수행하세요.
- on = 'None'으로 교집합을 만들어 출력하세요.
- 첫번째 df는 stock_name, 두번째 df는 name을 기준으로 합집합을(left)을 만들어 출력하세요.
- 오른쪽 df를 기준으로 합친 후 출력하세요.
- df1에서 가격이 50000 미만인 데이터만 출력하세요.
- df2에서 per가 30 보다 크고 pbr이 0.8 보다 큰 데이터를 모두 출력하세요.

In [179]:
df_s1 = pd.read_csv('dataset/stock_price.csv', encoding='cp949')
df_s2 = pd.read_csv('dataset/stock_valuation.csv', encoding='cp949')
print(df_s1,'\n')
print(df_s2)

       id stock_name          value   price
0  128940       한미약품   59385.666670  421000
1  130960     CJ E&M   58540.666670   98900
2  138250      엔에스쇼핑   14558.666670   13200
3  139480        이마트  239230.833300  254500
4  142280     녹십자엠에스     468.833333   10200
5  145990        삼양사   82750.000000   82000
6  185750        종근당   40293.666670  100500
7  192400      쿠쿠홀딩스  179204.666700  177500
8  199800         툴젠   -2514.333333  115400
9  204210     모두투어리츠    3093.333333    3475 

       id       name           eps     bps        per       pbr
0  130960     CJ E&M   6301.333333   54068  15.695091  1.829178
1  136480         하림    274.166667    3551  11.489362  0.887074
2  138040    메리츠금융지주   2122.333333   14894   6.313806  0.899691
3  139480        이마트  18268.166670  295780  13.931338  0.860437
4  145990        삼양사   5741.000000  108090  14.283226  0.758627
5  161390      한국타이어   5648.500000   51341   7.453306  0.820007
6  181710  NHN엔터테인먼트   2110.166667   78434  30.755864  0.827447
7 

In [180]:
# on = 'None' 전체 컬럼을 가져와서
# on = 'None'으로 교집합을 만들어 출력하세요.
pd.merge(df_s1,df_s2,how='inner',on=None)

id stock_name          value   price    name           eps     bps  \
0  130960     CJ E&M   58540.666670   98900  CJ E&M   6301.333333   54068   
1  139480        이마트  239230.833300  254500     이마트  18268.166670  295780   
2  145990        삼양사   82750.000000   82000     삼양사   5741.000000  108090   
3  185750        종근당   40293.666670  100500     종근당   3990.333333   40684   
4  204210     모두투어리츠    3093.333333    3475  모두투어리츠     85.166667    5335   

         per       pbr  
0  15.695091  1.829178  
1  13.931338  0.860437  
2  14.283226  0.758627  
3  25.185866  2.470259  
4  40.802348  0.651359

In [198]:
# 첫번째 df는 stock_name, 두번째 df는 name을 기준으로 합집합을(left)을 만들어 출력하세요.

df1=pd.merge(df_s1,df_s2,how='left',left_on='stock_name',right_on='name')
df1

id_x stock_name          value   price      id_y    name           eps  \
0  128940       한미약품   59385.666670  421000       NaN     NaN           NaN   
1  130960     CJ E&M   58540.666670   98900  130960.0  CJ E&M   6301.333333   
2  138250      엔에스쇼핑   14558.666670   13200       NaN     NaN           NaN   
3  139480        이마트  239230.833300  254500  139480.0     이마트  18268.166670   
4  142280     녹십자엠에스     468.833333   10200       NaN     NaN           NaN   
5  145990        삼양사   82750.000000   82000  145990.0     삼양사   5741.000000   
6  185750        종근당   40293.666670  100500  185750.0     종근당   3990.333333   
7  192400      쿠쿠홀딩스  179204.666700  177500       NaN     NaN           NaN   
8  199800         툴젠   -2514.333333  115400       NaN     NaN           NaN   
9  204210     모두투어리츠    3093.333333    3475  204210.0  모두투어리츠     85.166667   

        bps        per       pbr  
0       NaN        NaN       NaN  
1   54068.0  15.695091  1.829178  
2       NaN        NaN       NaN  
3  295780.0  13.931338  0.860437  
4       NaN        NaN       NaN  
5  108090.0  14.283226  0.758627  
6   40684.0  25.185866  2.470259  
7       NaN        NaN       NaN  
8       NaN        NaN       NaN  
9    5335.0  40.802348  0.651359

In [182]:
# 오른쪽 df를 기준으로 합친 후 출력하세요.

df2=pd.merge(df_s1,df_s2,how='right',left_on='stock_name',right_on='name')
df2

id_x stock_name          value     price    id_y       name  \
0  130960.0     CJ E&M   58540.666670   98900.0  130960     CJ E&M   
1       NaN        NaN            NaN       NaN  136480         하림   
2       NaN        NaN            NaN       NaN  138040    메리츠금융지주   
3  139480.0        이마트  239230.833300  254500.0  139480        이마트   
4  145990.0        삼양사   82750.000000   82000.0  145990        삼양사   
5       NaN        NaN            NaN       NaN  161390      한국타이어   
6       NaN        NaN            NaN       NaN  181710  NHN엔터테인먼트   
7  185750.0        종근당   40293.666670  100500.0  185750        종근당   
8  204210.0     모두투어리츠    3093.333333    3475.0  204210     모두투어리츠   
9       NaN        NaN            NaN       NaN  207940   삼성바이오로직스   

            eps     bps        per       pbr  
0   6301.333333   54068  15.695091  1.829178  
1    274.166667    3551  11.489362  0.887074  
2   2122.333333   14894   6.313806  0.899691  
3  18268.166670  295780  13.931338  0.860437  
4   5741.000000  108090  14.283226  0.758627  
5   5648.500000   51341   7.453306  0.820007  
6   2110.166667   78434  30.755864  0.827447  
7   3990.333333   40684  25.185866  2.470259  
8     85.166667    5335  40.802348  0.651359  
9   4644.166667   60099  89.790059  6.938551

In [199]:
# df1에서 가격이 50000 미만인 데이터만 출력하세요.

df1.loc[(df1['price']<50000),:].head()

id_x stock_name         value  price      id_y    name        eps  \
2  138250      엔에스쇼핑  14558.666670  13200       NaN     NaN        NaN   
4  142280     녹십자엠에스    468.833333  10200       NaN     NaN        NaN   
9  204210     모두투어리츠   3093.333333   3475  204210.0  모두투어리츠  85.166667   

      bps        per       pbr  
2     NaN        NaN       NaN  
4     NaN        NaN       NaN  
9  5335.0  40.802348  0.651359

In [201]:
price = df1['price']<50000
df1 = df1.loc[price,:]
df1.head()

id_x stock_name         value  price      id_y    name        eps  \
2  138250      엔에스쇼핑  14558.666670  13200       NaN     NaN        NaN   
4  142280     녹십자엠에스    468.833333  10200       NaN     NaN        NaN   
9  204210     모두투어리츠   3093.333333   3475  204210.0  모두투어리츠  85.166667   

      bps        per       pbr  
2     NaN        NaN       NaN  
4     NaN        NaN       NaN  
9  5335.0  40.802348  0.651359

In [192]:
# df2에서 per가 30 보다 크고 pbr이 0.8 보다 큰 데이터를 모두 출력하세요.
df2.loc[(df2['per']>30)&(df2['pbr']>0.8),:].head()

id_x stock_name  value  price    id_y       name          eps    bps  \
6   NaN        NaN    NaN    NaN  181710  NHN엔터테인먼트  2110.166667  78434   
9   NaN        NaN    NaN    NaN  207940   삼성바이오로직스  4644.166667  60099   

         per       pbr  
6  30.755864  0.827447  
9  89.790059  6.938551

### 과제(7/4)  (Q16 ~ Q20)

Q16. 함수 매핑을 이용하여 age에 각각 10을 곱하고 5를 더한 후 다시 더한 결과 값이 출력되게 하세요.

In [1]:
# (a * b) + (a + c)
# 데이터 가져오기
import seaborn as sns

# titanic 데이터셋에서 

titanic = sns.load_dataset('titanic')
# print(titanic.head())
df = titanic.loc[:, ['age','fare']]
df.head()

age     fare
0  22.0   7.2500
1  38.0  71.2833
2  26.0   7.9250
3  35.0  53.1000
4  35.0   8.0500

In [3]:
def mul_add(a,b,c):
    return (a * b) +(a + c)

sr1 = df.age.apply(mul_add,b=10,c=5)
sr1

0      247.0
1      423.0
2      291.0
3      390.0
4      390.0
       ...  
886    302.0
887    214.0
888      NaN
889    291.0
890    357.0
Name: age, Length: 891, dtype: float64

Q17. titanic.fare 데이터를 소수 2째자리 형태로 모두 출력하세요.
- 7.2500 -> 7.25

In [103]:
df = df.round({'fare':2})
df.head()

age   fare
0  22.0   7.25
1  38.0  71.28
2  26.0   7.92
3  35.0  53.10
4  35.0   8.05

Q18. df에서 fare와 age를 선택하여 100을 곱한 후 출력하세요

In [104]:
import pandas as pd

titanic = sns.load_dataset('titanic')
df = titanic
df.head()

survived  pclass     sex   age  sibsp  parch     fare embarked  class  \
0         0       3    male  22.0      1      0   7.2500        S  Third   
1         1       1  female  38.0      1      0  71.2833        C  First   
2         1       3  female  26.0      0      0   7.9250        S  Third   
3         1       1  female  35.0      1      0  53.1000        S  First   
4         0       3    male  35.0      0      0   8.0500        S  Third   

     who  adult_male deck  embark_town alive  alone  
0    man        True  NaN  Southampton    no  False  
1  woman       False    C    Cherbourg   yes  False  
2  woman       False  NaN  Southampton   yes   True  
3  woman       False    C  Southampton   yes  False  
4    man        True  NaN  Southampton    no   True

In [105]:
def mul_100(a):
    return a * 100

df[['age','fare']].applymap(mul_100).head()

age     fare
0  2200.0   725.00
1  3800.0  7128.33
2  2600.0   792.50
3  3500.0  5310.00
4  3500.0   805.00

Q19. join() 메소드는 두 데이터프레임의 행 인덱스를 기준으로 결합한다. 2개의 주식데이터를 가져와서 join() 메소드로 아래와 같이 결합한 후 다음 사항을 수행하세요.
- df1과 df2의 교집합만 출력되도록 결합하여 df4에 저장하고 출력
- df4에서 중복된 칼럼을 삭제한 후 블린 인덱싱을 이용하여 eps가 3000 보다 적거나 stock_name이 이마트인 데이터를 선택하여 데이터프레임을 생성하고 df5 이름으로 저장 및 출력하세요.(단, '<' 와 '==' 를 반드시 사용해야 함)

In [106]:
import pandas as pd

# 주식 데이터를 가져와서 데이터프레임 만들기
df1 = pd.read_excel('./dataset/stock price.xlsx', index_col='id')
df2 = pd.read_excel('./dataset/stock valuation.xlsx', index_col='id')

# 데이터프레임 결합(join)
# df1의 행 인덱스를 기준으로 결합하는 how='left'옵션이 기본 적용
df3 = df1.join(df2)
print(df3)


       stock_name          value   price    name           eps       bps  \
id                                                                         
128940       한미약품   59385.666667  421000     NaN           NaN       NaN   
130960     CJ E&M   58540.666667   98900  CJ E&M   6301.333333   54068.0   
138250      엔에스쇼핑   14558.666667   13200     NaN           NaN       NaN   
139480        이마트  239230.833333  254500     이마트  18268.166667  295780.0   
142280     녹십자엠에스     468.833333   10200     NaN           NaN       NaN   
145990        삼양사   82750.000000   82000     삼양사   5741.000000  108090.0   
185750        종근당   40293.666667  100500     종근당   3990.333333   40684.0   
192400      쿠쿠홀딩스  179204.666667  177500     NaN           NaN       NaN   
199800         툴젠   -2514.333333  115400     NaN           NaN       NaN   
204210     모두투어리츠    3093.333333    3475  모두투어리츠     85.166667    5335.0   

              per       pbr  
id                           
128940        NaN       NaN

In [107]:
df4 = df1.join(df2, how='inner')
df4

stock_name          value   price    name           eps     bps  \
id                                                                       
130960     CJ E&M   58540.666667   98900  CJ E&M   6301.333333   54068   
139480        이마트  239230.833333  254500     이마트  18268.166667  295780   
145990        삼양사   82750.000000   82000     삼양사   5741.000000  108090   
185750        종근당   40293.666667  100500     종근당   3990.333333   40684   
204210     모두투어리츠    3093.333333    3475  모두투어리츠     85.166667    5335   

              per       pbr  
id                           
130960  15.695091  1.829178  
139480  13.931338  0.860437  
145990  14.283226  0.758627  
185750  25.185866  2.470259  
204210  40.802348  0.651359

In [108]:
df4.T.duplicated()

stock_name    False
value         False
price         False
name           True
eps           False
bps           False
per           False
pbr           False
dtype: bool

In [109]:
~df4.T.duplicated()

stock_name     True
value          True
price          True
name          False
eps            True
bps            True
per            True
pbr            True
dtype: bool

In [110]:
df4.loc[:,~df4.T.duplicated()]

stock_name          value   price           eps     bps        per  \
id                                                                          
130960     CJ E&M   58540.666667   98900   6301.333333   54068  15.695091   
139480        이마트  239230.833333  254500  18268.166667  295780  13.931338   
145990        삼양사   82750.000000   82000   5741.000000  108090  14.283226   
185750        종근당   40293.666667  100500   3990.333333   40684  25.185866   
204210     모두투어리츠    3093.333333    3475     85.166667    5335  40.802348   

             pbr  
id                
130960  1.829178  
139480  0.860437  
145990  0.758627  
185750  2.470259  
204210  0.651359

In [111]:
df5= df4[(df4['eps']<3000)|(df4['stock_name']=='이마트')]
df5

stock_name          value   price    name           eps     bps  \
id                                                                       
139480        이마트  239230.833333  254500     이마트  18268.166667  295780   
204210     모두투어리츠    3093.333333    3475  모두투어리츠     85.166667    5335   

              per       pbr  
id                           
139480  13.931338  0.860437  
204210  40.802348  0.651359

Q20. 아래 df에서 중복 행을 제거한 후 df2에 저장하고 출력하세요.


In [46]:
import pandas as pd
df = pd.DataFrame({'c1':['a', 'a', 'b', 'a', 'b'],
                  'c2':[1, 1, 1, 2, 2],
                  'c3':[1, 1, 2, 2, 2]})
print(df)


  c1  c2  c3
0  a   1   1
1  a   1   1
2  b   1   2
3  a   2   2
4  b   2   2


In [47]:
df2 = df.drop_duplicates()
df2

c1  c2  c3
0  a   1   1
2  b   1   2
3  a   2   2
4  b   2   2

### 과제(7/5) (Q21,Q33 ~ Q34)

Q21. Q20의 df에서 drop_duplicates() 메소드를 사용하여 c2, c3열을 기준으로 중복 행을 제거한 후 df3에 저장하고 출력하세요.

In [48]:
df3 = df.drop_duplicates(subset=['c2','c3'])
df3

c1  c2  c3
0  a   1   1
2  b   1   2
3  a   2   2

Q33. './dataset/stock-data.csv'를 데이터프레임으로 불러와서 datetime64 자료형으로 변환한 후에 년, 월, 일로 분리하고 year를 인덱스로 셋팅하여 출력하세요.

In [65]:
df = pd.read_csv('dataset/stock-data.csv')
df.head()

Date  Close  Start   High    Low  Volume
0  2018-07-02  10100  10850  10900  10000  137977
1  2018-06-29  10700  10550  10900   9990  170253
2  2018-06-28  10400  10900  10950  10150  155769
3  2018-06-27  10900  10800  11050  10500  133548
4  2018-06-26  10800  10900  11000  10700   63039

In [66]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20 entries, 0 to 19
Data columns (total 6 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   Date    20 non-null     object
 1   Close   20 non-null     int64 
 2   Start   20 non-null     int64 
 3   High    20 non-null     int64 
 4   Low     20 non-null     int64 
 5   Volume  20 non-null     int64 
dtypes: int64(5), object(1)
memory usage: 1.1+ KB


In [67]:
# datetime64 자료형으로 변환
df['Date'] = pd.to_datetime(df['Date'])
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20 entries, 0 to 19
Data columns (total 6 columns):
 #   Column  Non-Null Count  Dtype         
---  ------  --------------  -----         
 0   Date    20 non-null     datetime64[ns]
 1   Close   20 non-null     int64         
 2   Start   20 non-null     int64         
 3   High    20 non-null     int64         
 4   Low     20 non-null     int64         
 5   Volume  20 non-null     int64         
dtypes: datetime64[ns](1), int64(5)
memory usage: 1.1 KB


In [68]:
# 년, 월, 일로 분리

df['year'] = df['Date'].dt.year
df['month'] = df['Date'].dt.month
df['day'] = df['Date'].dt.day
df.head()

Date  Close  Start   High    Low  Volume  year  month  day
0 2018-07-02  10100  10850  10900  10000  137977  2018      7    2
1 2018-06-29  10700  10550  10900   9990  170253  2018      6   29
2 2018-06-28  10400  10900  10950  10150  155769  2018      6   28
3 2018-06-27  10900  10800  11050  10500  133548  2018      6   27
4 2018-06-26  10800  10900  11000  10700   63039  2018      6   26

In [69]:
# year를 인덱스로 셋팅하여 출력하세요.
df1=df.copy()
df1.set_index('year', inplace=True)
df1.head()

Date  Close  Start   High    Low  Volume  month  day
year                                                           
2018 2018-07-02  10100  10850  10900  10000  137977      7    2
2018 2018-06-29  10700  10550  10900   9990  170253      6   29
2018 2018-06-28  10400  10900  10950  10150  155769      6   28
2018 2018-06-27  10900  10800  11050  10500  133548      6   27
2018 2018-06-26  10800  10900  11000  10700   63039      6   26

Q34. './dataset/stock-data.csv'를 데이터프레임으로 불러와서 datetime64 자료형으로 변환한 후에 Period로 변환하고 년월일 표기 변경하여 년월을 인덱스로 셋팅하여 출력하세요.


In [19]:
df = pd.read_csv('dataset/stock-data.csv')
df.head()

Date  Close  Start   High    Low  Volume
0  2018-07-02  10100  10850  10900  10000  137977
1  2018-06-29  10700  10550  10900   9990  170253
2  2018-06-28  10400  10900  10950  10150  155769
3  2018-06-27  10900  10800  11050  10500  133548
4  2018-06-26  10800  10900  11000  10700   63039

In [20]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20 entries, 0 to 19
Data columns (total 6 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   Date    20 non-null     object
 1   Close   20 non-null     int64 
 2   Start   20 non-null     int64 
 3   High    20 non-null     int64 
 4   Low     20 non-null     int64 
 5   Volume  20 non-null     int64 
dtypes: int64(5), object(1)
memory usage: 1.1+ KB


In [21]:
# datetime64 자료형으로 변환
df['Date'] = pd.to_datetime(df['Date'])
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20 entries, 0 to 19
Data columns (total 6 columns):
 #   Column  Non-Null Count  Dtype         
---  ------  --------------  -----         
 0   Date    20 non-null     datetime64[ns]
 1   Close   20 non-null     int64         
 2   Start   20 non-null     int64         
 3   High    20 non-null     int64         
 4   Low     20 non-null     int64         
 5   Volume  20 non-null     int64         
dtypes: datetime64[ns](1), int64(5)
memory usage: 1.1 KB


In [14]:
df

Date  Close  Start   High    Low  Volume
0  2018-07-02  10100  10850  10900  10000  137977
1  2018-06-29  10700  10550  10900   9990  170253
2  2018-06-28  10400  10900  10950  10150  155769
3  2018-06-27  10900  10800  11050  10500  133548
4  2018-06-26  10800  10900  11000  10700   63039
5  2018-06-25  11150  11400  11450  11000   55519
6  2018-06-22  11300  11250  11450  10750  134805
7  2018-06-21  11200  11350  11750  11200  133002
8  2018-06-20  11550  11200  11600  10900  308596
9  2018-06-19  11300  11850  11950  11300  180656
10 2018-06-18  12000  13400  13400  12000  309787
11 2018-06-15  13400  13600  13600  12900  201376
12 2018-06-14  13450  13200  13700  13150  347451
13 2018-06-12  13200  12200  13300  12050  558148
14 2018-06-11  11950  12000  12250  11950   62293
15 2018-06-08  11950  11950  12200  11800   59258
16 2018-06-07  11950  12200  12300  11900   49088
17 2018-06-05  12150  11800  12250  11800   42485
18 2018-06-04  11900  11900  12200  11700   25171
19 2018-06-01  11900  11800  12100  11750   32062

In [15]:
df1=df.copy()
df1.set_index('Date', inplace=True)
df1.head()

Close  Start   High    Low  Volume
Date                                          
2018-07-02  10100  10850  10900  10000  137977
2018-06-29  10700  10550  10900   9990  170253
2018-06-28  10400  10900  10950  10150  155769
2018-06-27  10900  10800  11050  10500  133548
2018-06-26  10800  10900  11000  10700   63039

In [16]:
# Period로 변환하고 년월일 표기 변경하여 년월을 인덱스로 셋팅하여 출력하세요.
df1.index = df1.index.to_period(freq='M')
df1.index

PeriodIndex(['2018-07', '2018-06', '2018-06', '2018-06', '2018-06', '2018-06',
             '2018-06', '2018-06', '2018-06', '2018-06', '2018-06', '2018-06',
             '2018-06', '2018-06', '2018-06', '2018-06', '2018-06', '2018-06',
             '2018-06', '2018-06'],
            dtype='period[M]', name='Date')

In [17]:
df1.head()

Close  Start   High    Low  Volume
Date                                       
2018-07  10100  10850  10900  10000  137977
2018-06  10700  10550  10900   9990  170253
2018-06  10400  10900  10950  10150  155769
2018-06  10900  10800  11050  10500  133548
2018-06  10800  10900  11000  10700   63039

In [24]:
df1=df.copy()
df1['year_mon'] = df1['Date'].dt.to_period(freq = 'M')
df1.set_index('year_mon', inplace=True)
df1.head()

Date  Close  Start   High    Low  Volume
year_mon                                               
2018-07  2018-07-02  10100  10850  10900  10000  137977
2018-06  2018-06-29  10700  10550  10900   9990  170253
2018-06  2018-06-28  10400  10900  10950  10150  155769
2018-06  2018-06-27  10900  10800  11050  10500  133548
2018-06  2018-06-26  10800  10900  11000  10700   63039

## 인덱스 재정렬 (reindex)

In [4]:
import numpy as np
import pandas as pd

In [5]:
col1 = np.random.uniform(10,20,size=10) # 10 ~ 20 사이 실수 10개
col2 = np.random.uniform(10,20,size=10)
sr1 = pd.Series(col1,name='col1')
sr2 = pd.Series(col2,name='col2')
df=pd.concat([sr1,sr2],axis=1)
df

col1       col2
0  12.366868  18.803923
1  11.022081  10.180513
2  17.941459  15.064400
3  18.115248  10.428434
4  14.095500  16.307345
5  18.500918  12.500308
6  18.082954  17.160042
7  10.401931  10.228704
8  18.351371  13.876543
9  18.984224  11.611956

### round()

In [6]:
df = df.round({'col1':2,'col2':1}) # 컬럼 별 소수점 지정
df.head()

col1  col2
0  12.37  18.8
1  11.02  10.2
2  17.94  15.1
3  18.12  10.4
4  14.10  16.3

### astype()

In [53]:
# col2를 정수로 변환하여 출력
df1 = df.copy()
df1 = df1.col2.astype(int) # 정수 type으로 변경
df1

0    11
1    18
2    11
3    19
4    13
5    12
6    15
7    11
8    15
9    13
Name: col2, dtype: int32

In [54]:
df2 = df.copy()
df2.iloc[1:3,:] = np.nan
df2.head()

col1  col2
0  15.80  11.1
1    NaN   NaN
2    NaN   NaN
3  16.83  19.1
4  13.34  13.3

In [59]:
# Null 값을 0으로 채우기
df2.fillna(0,inplace=True)

# 정수로 변경
df2 = df2.astype({'col1':'int','col2':'int'})
df2.head()

col1  col2
0    15    11
1     0     0
2     0     0
3    16    19
4    13    13

### set_index()

In [61]:
# 칼럼 추가
df2['col0']=list('abcdefghij')

# 컬럼을 인덱스로 바꿀 때
df2.set_index('col0', inplace = True) 
df2.head()

col1  col2
col0            
a       15    11
b        0     0
c        0     0
d       16    19
e       13    13

In [62]:
# null값 확인
df2.isna().sum()

col1    0
col2    0
dtype: int64

### reset_index()

In [63]:
# 인덱스 초기화 함수 , col0 인덱스를 컬럼으로
# 기존 인덱스 삭제하려면 drop = True, 디폴트는 False
df2 = df2.reset_index()
df2.head()

col0  col1  col2
0    a    15    11
1    b     0     0
2    c     0     0
3    d    16    19
4    e    13    13

### reindex()

In [64]:
# 인덱스의 재배치
# index 내가 원하는 인덱스만 가지고 올 수 있다.
df3 = df2.reindex(index=[1,2],columns=['col1','col2'])
df3

col1  col2
1     0     0
2     0     0

In [65]:
df4 = df2.reindex(index=[3,4], columns=['col1','col2'])
df4

col1  col2
3    16    19
4    13    13

## 데이터 스케일링 (Scaling )
- MinMaxScaler, StandarScaler

### MinMaxScaler

In [69]:
from sklearn.datasets import load_boston
import warnings
warnings.filterwarnings('ignore')
boston = load_boston()
print(boston.keys())
print(boston.feature_names)
df_boston = pd.DataFrame(boston.data, columns = boston.feature_names)
df_boston['target'] = boston.target
df_boston.head()

dict_keys(['data', 'target', 'feature_names', 'DESCR', 'filename', 'data_module'])
['CRIM' 'ZN' 'INDUS' 'CHAS' 'NOX' 'RM' 'AGE' 'DIS' 'RAD' 'TAX' 'PTRATIO'
 'B' 'LSTAT']


CRIM    ZN  INDUS  CHAS    NOX     RM   AGE     DIS  RAD    TAX  \
0  0.00632  18.0   2.31   0.0  0.538  6.575  65.2  4.0900  1.0  296.0   
1  0.02731   0.0   7.07   0.0  0.469  6.421  78.9  4.9671  2.0  242.0   
2  0.02729   0.0   7.07   0.0  0.469  7.185  61.1  4.9671  2.0  242.0   
3  0.03237   0.0   2.18   0.0  0.458  6.998  45.8  6.0622  3.0  222.0   
4  0.06905   0.0   2.18   0.0  0.458  7.147  54.2  6.0622  3.0  222.0   

   PTRATIO       B  LSTAT  target  
0     15.3  396.90   4.98    24.0  
1     17.8  396.90   9.14    21.6  
2     17.8  392.83   4.03    34.7  
3     18.7  394.63   2.94    33.4  
4     18.7  396.90   5.33    36.2

In [ ]:
# sklearn 라이브러리의 스케일러 MinMaxScaler를 사용한 스케일링
스케일러의 fit_transform() 사용시 2차원의 데이터를 전달해야 함
(DataFrame도 2차원), 결과는 ndarray로 반환됨

In [ ]:
# transform() 메소드는 그룹별로 함수를 적용하지만 본래 행 인덱스와 열 이름을 기준으로 연산 결과를 반환한다.

In [72]:
from sklearn.preprocessing import MinMaxScaler

scaler = MinMaxScaler()
df_boston['RM']=scaler.fit_transform(df_boston[['RM']])
df_boston[['RM']].describe()

RM
count  506.000000
mean     0.521869
std      0.134627
min      0.000000
25%      0.445392
50%      0.507281
75%      0.586798
max      1.000000

### StandardScaler

In [ ]:
# sklearn 라이브러리의 스케일러 StandardScaler를 사용한 스케일링

In [73]:
from sklearn.preprocessing import StandardScaler
scaler = StandardScaler()
df_boston['INDUS'] = scaler.fit_transform(df_boston[['INDUS']])
df_boston[['INDUS']].describe()

INDUS
count  5.060000e+02
mean   2.804081e-16
std    1.000990e+00
min   -1.557842e+00
25%   -8.676906e-01
50%   -2.110985e-01
75%    1.015999e+00
max    2.422565e+00

## 데이터 다루기 (encoding)

In [3]:
import pandas as pd
import numpy as np
import seaborn as sns

titanic = sns.load_dataset('titanic')
# 특정 컬럼의 값을 변경
titanic['sex'] = titanic.sex.replace({'male':0,'female':1})
# 열이름 변경
titanic.rename(columns={'sex':'gender','age':'age_cat','class':'c_level'},inplace=True)
# 컬럼명을 대문자로 변경
titanic.columns = map(lambda x:str(x).upper(), titanic.columns)
# 컬럼명을 소문자로 변경
titanic.rename(columns=str.lower,inplace=True)
df = titanic[['survived','pclass','gender','c_level']]
df.head()



survived  pclass  gender c_level
0         0       3       0   Third
1         1       1       1   First
2         1       3       1   Third
3         1       1       1   First
4         0       3       0   Third

In [13]:
titanic.columns

Index(['survived', 'pclass', 'gender', 'age_cat', 'sibsp', 'parch', 'fare',
       'embarked', 'c_level', 'who', 'adult_male', 'deck', 'embark_town',
       'alive', 'alone'],
      dtype='object')

In [33]:
import warnings
warnings.filterwarnings('ignore')
df['gender'].replace([1,0],['female','male'],inplace = True)
df.head()

survived  pclass  gender c_level
0         0       3    male   Third
1         1       1  female   First
2         1       3  female   Third
3         1       1  female   First
4         0       3    male   Third

### LabelEncoder

In [34]:
# 인코딩 - LabelEncoder
from sklearn.preprocessing import LabelEncoder

le = LabelEncoder()
features = ['gender','c_level']
for f in features:
    df1[f] = le.fit_transform(df1[f])
df1.head()

survived  pclass  gender  c_level
0         0       3       1        2
1         1       1       0        0
2         1       3       0        2
3         1       1       0        0
4         0       3       1        2

In [35]:
df1.c_level.value_counts()

2    491
0    216
1    184
Name: c_level, dtype: int64

### OneHotEncoder

In [38]:
df2=df.copy()
df3 = df2[['gender','c_level']]
pd.get_dummies(df3).head()

gender_female  gender_male  c_level_First  c_level_Second  c_level_Third
0              0            1              0               0              1
1              1            0              1               0              0
2              1            0              0               0              1
3              1            0              1               0              0
4              0            1              0               0              1

### 과제(7/5)

- titanic 데이터셋에서 embarked, who, adult_male 에 대하여 LabelEncoding 하세요.
- titanic 데이터셋에서 embarked, who, adult_male 에 대하여 OneHotEncoding 하세요.

In [4]:
titanic = sns.load_dataset('titanic')
titanic.head()

survived  pclass     sex   age  sibsp  parch     fare embarked  class  \
0         0       3    male  22.0      1      0   7.2500        S  Third   
1         1       1  female  38.0      1      0  71.2833        C  First   
2         1       3  female  26.0      0      0   7.9250        S  Third   
3         1       1  female  35.0      1      0  53.1000        S  First   
4         0       3    male  35.0      0      0   8.0500        S  Third   

     who  adult_male deck  embark_town alive  alone  
0    man        True  NaN  Southampton    no  False  
1  woman       False    C    Cherbourg   yes  False  
2  woman       False  NaN  Southampton   yes   True  
3  woman       False    C  Southampton   yes  False  
4    man        True  NaN  Southampton    no   True

In [77]:
from sklearn.preprocessing import LabelEncoder

le = LabelEncoder()
features = ['embarked', 'who', 'adult_male']
for f in features:
    titanic[f] = le.fit_transform(titanic[f])
titanic.head()

survived  pclass     sex   age  sibsp  parch     fare  embarked  class  \
0         0       3    male  22.0      1      0   7.2500         2  Third   
1         1       1  female  38.0      1      0  71.2833         0  First   
2         1       3  female  26.0      0      0   7.9250         2  Third   
3         1       1  female  35.0      1      0  53.1000         2  First   
4         0       3    male  35.0      0      0   8.0500         2  Third   

   who  adult_male deck  embark_town alive  alone  
0    1           1  NaN  Southampton    no  False  
1    2           0    C    Cherbourg   yes  False  
2    2           0  NaN  Southampton   yes   True  
3    2           0    C  Southampton   yes  False  
4    1           1  NaN  Southampton    no   True

In [79]:
titanic.embarked.value_counts()

2    644
0    168
1     77
3      2
Name: embarked, dtype: int64

In [80]:
titanic.who.value_counts()

1    537
2    271
0     83
Name: who, dtype: int64

In [81]:
titanic.adult_male.value_counts()

1    537
0    354
Name: adult_male, dtype: int64

In [ ]:
# OneHotEncoding

In [7]:
titanic = sns.load_dataset('titanic')
titanic.head()

survived  pclass     sex   age  sibsp  parch     fare embarked  class  \
0         0       3    male  22.0      1      0   7.2500        S  Third   
1         1       1  female  38.0      1      0  71.2833        C  First   
2         1       3  female  26.0      0      0   7.9250        S  Third   
3         1       1  female  35.0      1      0  53.1000        S  First   
4         0       3    male  35.0      0      0   8.0500        S  Third   

     who  adult_male deck  embark_town alive  alone  
0    man        True  NaN  Southampton    no  False  
1  woman       False    C    Cherbourg   yes  False  
2  woman       False  NaN  Southampton   yes   True  
3  woman       False    C  Southampton   yes  False  
4    man        True  NaN  Southampton    no   True

In [9]:
titanic.adult_male = titanic.adult_male.astype('str')
titanic1 = titanic[['embarked', 'who', 'adult_male']]
pd.get_dummies(titanic1).head()

embarked_C  embarked_Q  embarked_S  who_child  who_man  who_woman  \
0           0           0           1          0        1          0   
1           1           0           0          0        0          1   
2           0           0           1          0        0          1   
3           0           0           1          0        0          1   
4           0           0           1          0        1          0   

   adult_male_False  adult_male_True  
0                 0                1  
1                 1                0  
2                 1                0  
3                 1                0  
4                 0                1